internal/ingest/linksystem.go

```diff
@@ -147,67 +149,122 @@ func verifyAdvertisement(n ipld.Node) (schema.Advertisement, error) {
 			log.Errorf("Error while fetching the node from datastore: %s", err)
 			return
 		}
 
-		// Decode entries into an IPLD node
-		nentries, err := decodeIPLDNode(bytes.NewBuffer(val))
+		// Decode block to IPLD node
+		node, err := decodeIPLDNode(bytes.NewBuffer(val))
 		if err != nil {
 			log.Errorf("Error decoding ipldNode: %s", err)
 			return
 		}
 
-		// If this is an advertisement, then nothing to do yet.  Wait for the
-		// list of CIDs to ingest.
-		if isAdvertisement(nentries) {
+		// If this is an advertisement, sync entries within it.
+		if isAdvertisement(node) {
+			log.Debug("Incoming block is an advertisement")
+			ad, err := decodeAd(node)
+			if err != nil {
+				log.Errorf("Error decoding advertosement: %s", err)
+				return
+			}
+			log.Debug("Syncing entries")
+			// TODO: consider exposing config value for maximum sync timeout then set in context.
+			err = li.syncAdEntries(context.TODO(), p, ad)
+			if err != nil {
+				log.Errorf("Error syncing advertosement entries: %s", err)
+			}
 			return
 		}
 
 		// Get the advertisement ID corresponding to the link.
 		// From the reverse map.
-		log.Debug("hook - Not an advertisement, let's start ingesting entries")
+		log.Debug("Incoming block is not an advertisement; processing entries")
 		adCid, err := getCidToAdMapping(li.ds, c)
 		if err != nil {
 			log.Errorf("Error getting advertisementID for entries map: %s", err)
 			return
 		}
+		log = log.With("adCid", adCid)
 
-		log.Infow("hook - Processing entries", "ad", adCid, "link", c)
-		// Process entries and ingest them.
-		err = li.processEntries(adCid, p, nentries)
+		log.Infow("Processing entries")
+		err = li.processEntries(adCid, p, node)
 		if err != nil {
-			log.Errorf("Error processing entries for advertisement: %s", err)
+			log.Errorw("Error processing entries for advertisement", "err", err)
 			return
 		}
 
 		li.sigUpdate <- struct{}{}
 
 		// Remove the datastore entry that maps a chunk to an advertisement
 		// now that the chunk is processed.
+		log.Debug("Removing mapping to advertisement for processed entries")
 		err = deleteCidToAdMapping(li.ds, c)
 		if err != nil {
-			log.Errorf("Error deleting cid-advertisement mapping for entries: %s", err)
+			log.Errorw("Error deleting cid-advertisement mapping for entries", "err", err)
 		}
 
-		log.Debug("hook - Removing processed entries from datastore")
 		// Remove the index from the data store now that processing it has
 		// finished.  This prevents storing redundant information in
 		// several datastore.
+		log.Debug("Removing processed entries from datastore")
 		err = li.ds.Delete(dsKey(c.String()))
 		if err != nil {
-			log.Errorf("Error deleting index from datastore: %s", err)
+			log.Errorw("Error deleting index from datastore", "err", err)
 		}
 	}
 }
 
+func (li *legIngester) syncAdEntries(ctx context.Context, from peer.ID, ad schema.Advertisement) error {
+	log := log.With("peerID", from)
+	sub, err := li.newPeerSubscriber(ctx, from)
```

Comment:
After creating this subscriber to sync the ad entries, does it need to be shutdown when the sync is done?

Reply:
After the legs broker stuff is available, we will just call Sync() here and not have to create a subscriber at all.  The broker will handle using the existing handler (or creating one if needed).

Reply:
I think the name `newPeerSubscriber` is misleading; it actually gets or creates a new subscriber and there is long-lived subscriber map in ingest engine. See existing implementation [here](https://github.com/filecoin-project/storetheindex/blob/main/internal/ingest/ingest.go#L267).